### ISSUE TITLE
Should Curio Support IOCP?

### ISSUE THREAD

**Issue by dabeaz** [dabeaz/curio#68]
If Curio is to support Windows, does it do so using IOCP?  If so, what does it look like?  Is it just an internal implementation detail hidden from view?  Or would programmers want to interact with it more directly in some manner?   

Just thinking out loud here....

**Comment by gigatexal** [comment]
It's been in Windows since forever so it'll let curio run on any version of Windows. So that's cool.

**Comment by Lukasa** [comment]
IOCP is cool, because it's not the same as select and friends at all. Rather than being a reactor kind of model, it's much more a proactor kind of model. Very neat. Well worth supporting if you can make it happen.

**Comment by AndreLouisCaron** [comment]
If you're going to support Windows, it should most definitely support IOCP.  The Win32 implementation of [select()](https://msdn.microsoft.com/library/ms740141.aspx) has several important limitations:

- only works for sockets (no pipes, no thread joins, no process joins, etc.)
- can handle a maximum of 64 file descriptors per fdset

IMO, curio on Windows without IOCP is not curio :-) 

I'm currently looking into this.  I think the proactor VS reactor duality is going to have some non-trivial impacts on the code if we want proper multi-platform support and I'm not familiar enough with curio internals to clearly see how it's going to end up.  I was thinking of putting together an IOCP demo using the most minimalistic/crude curio-like kernel in order to sketch out what is required.

I'll keep you posted as soon as I get the demo up and running and we'll see if/how we proceed from there :-)

**Comment by njsmith** [comment]
The Curio's public API is actually much closer to IOCP than it is to Unix (e.g. in IOCP, `sock.sendall` is a primitive), so I suspect that the way to minimize impedence mismatch is to create an abstraction layer that looks like IOCP, implement curio's public API on top fo that, and then implement the abstraction layer using IOCP on Windows and selectors on Unix-likes.

**Comment by njsmith** [comment]
Probably the simplest approach to doing this in curio would be for operations like `sendall` to become first-class traps. The alternative would be for the set of traps to vary on different systems and code like `sendall` would have to decide which traps to call depending on what OS it's running on, which seems like it might be more awkward.

Also FYI in case it's useful: asyncio has a Python wrapper for IOCP inside it. In the source it's `overlapped.c`; I don't have a windows machine handy so I'm not sure where it ends up (maybe `import asyncio.overlapped`?). I don't think it's considered public API, but for a proof-of-concept or as a starting point then it's probably quicker than writing a new wrapper from scratch :-). (In the long run I guess we probably would prefer a cffi wrapper for PyPy-friendliness, but that's low priority.)

**Comment by dabeaz** [comment]
Here's my take on this. The most important part of Curio is not the specific implementation of Curio, but the fact that it views async/await as the API.   Yes, there is an underlying kernel, but applications built with Curio do absolutely everything through async/await.  They don't hold a kernel reference.  They don't execute methods on the kernel.  The kernel is out of the picture.

This is important for exactly the kind of problem posed here.  I think you should be able to swap in a different kernel implementation (and maybe a modified implementation of the I/O wrappers for sockets and files) and still have applications work just fine.    As an example, I have a highly experimental Curio kernel written entirely in C that I've been fooling around with.   It can run Curio applications without modification---it's purely a matter of calling a different `run()` function to start things out.    I don't see any fundamental reason why an IOCP enabled kernel implementation couldn't work this way.

I think the most tricky part of this is simply where you put the I/O implementation.  Does it go in the kernel itself?  Does it sit outside inside a `Socket` class?  Maybe it's a mix of both. From an application point of view, I don't think it much matters--users aren't supposed to be writing low-level code that concerns itself with these details (e.g., if you call `await sock.recv(1000)` you don't really care how that works as long as it does). 

Big picture:  These are the key features you need to focus on:

1. Waiting for (or completing) I/O  
2. Waiting on Futures (for interaction with Threads/Processes)
3. Time (sleeping)

Everything else would likely be unchanged (i.e., spawning tasks, waiting on queues, etc.).

**Comment by AndreLouisCaron** [comment]
Wow, thanks for the quick replies :-)

@dabeaz  Yes, I'll be focusing on those 3 things for the demo:

1. seems to be all about `GetQueuedCompletionStatus()` and mapping the result back to the task that initiated the I/O operation.
3. I guess that using the timeout on the call to `GetQueuedCompletionStatus()` provides the minimum requirement when paired with heapq or something similar?
2. `PostQueuedCompletionStatus()` can be used to post a "fake" completion status to unblock `GetQueuedCompletionStatus()` from any other thread when necessary.

@njsmith I already started writing my own bindings since I know the IOCP API pretty well, but I'll take a look at what's already available in asyncio.

**Comment by njsmith** [comment]
These notes on Windows I/O APIs are probably of interest here too: https://github.com/python-trio/trio/issues/52
